File: leetcode/dp/2560_HouseRobberⅣ.py
from ..import *
# 贪心的解法
class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        l,r = min(nums),max(nums)+1
        def check(mx:int)->int:
            this_k = i = 0
            while i < len(nums):
                if nums[i] <= mx: this_k,i = this_k+1,i+2
                else: i += 1
            return this_k 
        while l < r:
            mid = l + (r - l) // 2
            if check(mid) >= k:
                r = mid
            else:
                l = mid + 1
        return r

File: DP/Leetcode/2560_HouseRobberⅣ.cpp
#include <bits/stdc++.h>

using namespace std;
// dp解法
class Solution {
public:
    bool check(vector<int> nums,int k, int mx) {
        int interval = 0, neighbor = 0;
        for (int n : nums) {
            if (n > mx)
                interval = neighbor;
            else {
                int tmp = neighbor;
                neighbor = max(neighbor, interval + 1);
                interval = tmp;
            }
        }
        return neighbor >= k;
    }
    int minCapability(vector<int> &nums, int k) {
        int l = 0, r = *max_element(nums.begin(), nums.end());
        while (l + 1 < r) { // 开区间写法
            int mid = l + (r - l) / 2;
            (check(nums,k,mid) ? r : l) = mid;
        }
        return r;
    }
};
int main() {

    //system("pause");
    return 0;
}
